File: message-security-layer/src/test/javascript/crypto/MslSignatureEnvelopeTest.js
/**
 * Copyright (c) 2013-2014 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * MSL signature envelope unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("MslSignatureEnvelope", function() {
    /** JSON key version. */
    var KEY_VERSION = "version";
    /** JSON key algorithm. */
    var KEY_ALGORITHM = "algorithm";
    /** JSON key signature. */
    var KEY_SIGNATURE = "signature";
    
    // Shortcuts.
    var Version = MslSignatureEnvelope$Version;
    
    var random = new Random();
    var SIGNATURE = new Uint8Array(32);
    random.nextBytes(SIGNATURE);

    describe("version 1", function() {
        it("ctors", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);
            
            var envelopeBytes, joEnvelope;
            runs(function() {
                expect(envelope.algorithm).toBeNull();
                expect(envelope.signature).toEqual(SIGNATURE);
                envelopeBytes = envelope.bytes;
                expect(envelopeBytes).not.toBeNull();

                MslSignatureEnvelope$parse(envelopeBytes, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);

            runs(function() {
                expect(joEnvelope.algorithm).toEqual(envelope.algorithm);
                expect(joEnvelope.signature).toEqual(envelope.signature);
                var joEnvelopeBytes = joEnvelope.bytes;
                expect(joEnvelopeBytes).toEqual(envelopeBytes);
            });
        });
        
        it("json is correct", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);
            
            runs(function() {
                var envelopeBytes = envelope.bytes;
                expect(envelopeBytes).not.toBeNull();
                expect(envelopeBytes).toEqual(SIGNATURE);
            });
        });
    });
    
    function data() {
        var keys = Object.keys(MslConstants$SignatureAlgo); 
        return keys.map(function(key) {
            return [ MslConstants$SignatureAlgo[key] ];
        });
    }
    
    parameterize("version 2", data, function(algorithm) {
        it("ctors", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);
            
            var envelopeBytes, joEnvelope;
            runs(function() {
                expect(envelope.algorithm).toEqual(algorithm);
                expect(envelope.signature).toEqual(SIGNATURE);
                envelopeBytes = envelope.bytes;
                expect(envelopeBytes).not.toBeNull();

                MslSignatureEnvelope$parse(envelopeBytes, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);
            
            runs(function() {
                expect(joEnvelope.algorithm).toEqual(envelope.algorithm);
                expect(joEnvelope.signature).toEqual(envelope.signature);
                var joEnvelopeBytes = joEnvelope.bytes;
                expect(joEnvelopeBytes).toEqual(envelopeBytes);
            });
        });
        
        it("json is correct", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);
            
            runs(function() {
                var envelopeBytes = envelope.bytes;
                var jo = JSON.parse(textEncoding$getString(envelopeBytes, MslConstants$DEFAULT_CHARSET));
                
                expect(parseInt(jo[KEY_VERSION])).toEqual(Version.V2);
                expect(jo[KEY_ALGORITHM]).toEqual(algorithm.toString());
                expect(base64$decode(jo[KEY_SIGNATURE])).toEqual(SIGNATURE);
            });
        });
        
        it("missing version", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);
            
            var joJson, joEnvelope;
            runs(function() {
                var envelopeBytes = envelope.bytes;
                var jo = JSON.parse(textEncoding$getString(envelopeBytes, MslConstants$DEFAULT_CHARSET));
                delete jo[KEY_VERSION];
                
                joJson = textEncoding$getBytes(JSON.stringify(jo), MslConstants$DEFAULT_CHARSET);
                MslSignatureEnvelope$parse(joJson, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);
            
            runs(function() {
                expect(joEnvelope.algorithm).toBeNull();
                expect(joEnvelope.signature).toEqual(joJson);
            });
        });
        
        it("invalid version", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);

            var joJson, joEnvelope;
            runs(function() {
                var envelopeBytes = envelope.bytes;
                var jo = JSON.parse(textEncoding$getString(envelopeBytes, MslConstants$DEFAULT_CHARSET));
                jo[KEY_VERSION] = "x";
                
                joJson = textEncoding$getBytes(JSON.stringify(jo), MslConstants$DEFAULT_CHARSET);
                MslSignatureEnvelope$parse(joJson, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);
            
            runs(function() {
                expect(joEnvelope.algorithm).toBeNull();
                expect(joEnvelope.signature).toEqual(joJson);
            });
        });
        
        it("unknown version", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);

            var joJson, joEnvelope;
            runs(function() {
                var envelopeBytes = envelope.bytes;
                var jo = JSON.parse(textEncoding$getString(envelopeBytes, MslConstants$DEFAULT_CHARSET));
                jo[KEY_VERSION] = "-1";

                joJson = textEncoding$getBytes(JSON.stringify(jo), MslConstants$DEFAULT_CHARSET);
                MslSignatureEnvelope$parse(joJson, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);
            
            runs(function() {
                expect(joEnvelope.algorithm).toBeNull();
                expect(joEnvelope.signature).toEqual(joJson);
            });
        });
        
        it("missing algorithm", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);

            var joJson, joEnvelope;
            runs(function() {
                var envelopeBytes = envelope.bytes;
                var jo = JSON.parse(textEncoding$getString(envelopeBytes, MslConstants$DEFAULT_CHARSET));
                delete jo[KEY_ALGORITHM];

                joJson = textEncoding$getBytes(JSON.stringify(jo), MslConstants$DEFAULT_CHARSET);
                MslSignatureEnvelope$parse(joJson, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);
            
            runs(function() {
                expect(joEnvelope.algorithm).toBeNull();
                expect(joEnvelope.signature).toEqual(joJson);
            });
        });
        
        it("invalid algorithm", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);

            var joJson, joEnvelope;
            runs(function() {
                var envelopeBytes = envelope.bytes;
                var jo = JSON.parse(textEncoding$getString(envelopeBytes, MslConstants$DEFAULT_CHARSET));
                jo[KEY_ALGORITHM] = "x";

                joJson = textEncoding$getBytes(JSON.stringify(jo), MslConstants$DEFAULT_CHARSET);
                MslSignatureEnvelope$parse(joJson, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);
            
            runs(function() {
                expect(joEnvelope.algorithm).toBeNull();
                expect(joEnvelope.signature).toEqual(joJson);
            });
        });
        
        it("missing signature", function() {
            var envelope;
            runs(function() {
                MslSignatureEnvelope$create(algorithm, SIGNATURE, {
                    result: function(x) { envelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return envelope; }, "envelope", 100);

            var joJson, joEnvelope;
            runs(function() {
                var envelopeBytes = envelope.bytes;
                var jo = JSON.parse(textEncoding$getString(envelopeBytes, MslConstants$DEFAULT_CHARSET));
                delete jo[KEY_SIGNATURE];

                joJson = textEncoding$getBytes(JSON.stringify(jo), MslConstants$DEFAULT_CHARSET);
                MslSignatureEnvelope$parse(joJson, null, {
                    result: function(x) { joEnvelope = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); }
                });
            });
            waitsFor(function() { return joEnvelope; }, "joEnvelope", 100);
            
            runs(function() {
                expect(joEnvelope.algorithm).toBeNull();
                expect(joEnvelope.signature).toEqual(joJson);
            });
        });
    });
});

File: tests/src/test/java/com/netflix/msl/crypto/MslSignatureEnvelopeTest.java
/**
 * Copyright (c) 2013-2014 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.netflix.msl.crypto;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Random;

import javax.xml.bind.DatatypeConverter;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import org.junit.runners.Suite;
import org.junit.runners.Suite.SuiteClasses;

import com.netflix.msl.MslConstants.SignatureAlgo;
import com.netflix.msl.MslCryptoException;
import com.netflix.msl.MslEncodingException;
import com.netflix.msl.crypto.MslSignatureEnvelope.Version;

/**
 * MSL signature envelope unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
@RunWith(Suite.class)
@SuiteClasses({MslSignatureEnvelopeTest.Version1.class,
               MslSignatureEnvelopeTest.Version2.class})
public class MslSignatureEnvelopeTest {
    /** JSON key version. */
    private final static String KEY_VERSION = "version";
    /** JSON key algorithm. */
    private final static String KEY_ALGORITHM = "algorithm";
    /** JSON key signature. */
    private final static String KEY_SIGNATURE = "signature";
    
    private static final byte[] SIGNATURE = new byte[32];
    
    @BeforeClass
    public static void setup() {
        final Random random = new Random();
        random.nextBytes(SIGNATURE);
    }
    
        public static class Version1 {
        @Test
        public void ctors() throws MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(SIGNATURE);
            assertNull(envelope.getAlgorithm());
            assertArrayEquals(SIGNATURE, envelope.getSignature());
            final byte[] envelopeBytes = envelope.getBytes();
            assertNotNull(envelopeBytes);
            
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(envelopeBytes);
            assertEquals(envelope.getAlgorithm(), joEnvelope.getAlgorithm());
            assertArrayEquals(envelope.getSignature(), joEnvelope.getSignature());
            final byte[] joEnvelopeBytes = joEnvelope.getBytes();
            assertArrayEquals(envelopeBytes, joEnvelopeBytes);
        }
        
        @Test
        public void json() throws MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(SIGNATURE);
            final byte[] envelopeBytes = envelope.getBytes();
            assertNotNull(envelopeBytes);
            assertArrayEquals(SIGNATURE, envelopeBytes);
        }
    }
    
    @RunWith(Parameterized.class)
    public static class Version2 {
        @Parameters
        public static Collection<Object[]> data() {
            final List<Object[]> params = new ArrayList<Object[]>();
            for (final SignatureAlgo algo: SignatureAlgo.values())
                params.add(new Object[] { algo });
            return params;
        }
        
        /** Algorithm. */
        private final SignatureAlgo algorithm;
        
        /**
         * Create a new Version 2 test set with the provided algorithm.
         * 
         * @param algorithm the algorithm.
         */
        public Version2(final SignatureAlgo algorithm) {
            this.algorithm = algorithm;
        }
        
        @Test
        public void ctors() throws MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);
            assertEquals(algorithm, envelope.getAlgorithm());
            assertArrayEquals(SIGNATURE, envelope.getSignature());
            final byte[] envelopeBytes = envelope.getBytes();
            assertNotNull(envelopeBytes);
            
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(envelopeBytes);
            assertEquals(envelope.getAlgorithm(), joEnvelope.getAlgorithm());
            assertArrayEquals(envelope.getSignature(), joEnvelope.getSignature());
            final byte[] joEnvelopeBytes = joEnvelope.getBytes();
            assertArrayEquals(envelopeBytes, joEnvelopeBytes);
        }
        
        @Test
        public void json() throws JSONException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);
            final byte[] envelopeBytes = envelope.getBytes();
            final JSONObject jo = new JSONObject(new String(envelopeBytes));
            
            assertEquals(Version.V2.intValue(), jo.getInt(KEY_VERSION));
            assertEquals(algorithm.toString(), jo.getString(KEY_ALGORITHM));
            assertArrayEquals(SIGNATURE, DatatypeConverter.parseBase64Binary(jo.getString(KEY_SIGNATURE)));
        }
        
        @Test
        public void missingVersion() throws JSONException, MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);

            final byte[] envelopeBytes = envelope.getBytes();
            final JSONObject jo = new JSONObject(new String(envelopeBytes));
            jo.remove(KEY_VERSION);
            
            final byte[] joJson = jo.toString().getBytes();
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(joJson);
            assertNull(joEnvelope.getAlgorithm());
            assertArrayEquals(joJson, joEnvelope.getSignature());
        }
        
        @Test
        public void invalidVersion() throws JSONException, MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);

            final byte[] envelopeBytes = envelope.getBytes();
            final JSONObject jo = new JSONObject(new String(envelopeBytes));
            jo.put(KEY_VERSION, "x");
            
            final byte[] joJson = jo.toString().getBytes();
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(joJson);
            assertNull(joEnvelope.getAlgorithm());
            assertArrayEquals(joJson, joEnvelope.getSignature());
        }
        
        @Test
        public void unknownVersion() throws JSONException, MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);

            final byte[] envelopeBytes = envelope.getBytes();
            final JSONObject jo = new JSONObject(new String(envelopeBytes));
            jo.put(KEY_VERSION, "-1");
            
            final byte[] joJson = jo.toString().getBytes();
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(joJson);
            assertNull(joEnvelope.getAlgorithm());
            assertArrayEquals(joJson, joEnvelope.getSignature());
        }
        
        @Test
        public void missingAlgorithm() throws JSONException, MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);

            final byte[] envelopeBytes = envelope.getBytes();
            final JSONObject jo = new JSONObject(new String(envelopeBytes));
            jo.remove(KEY_ALGORITHM);
            
            final byte[] joJson = jo.toString().getBytes();
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(joJson);
            assertNull(joEnvelope.getAlgorithm());
            assertArrayEquals(joJson, joEnvelope.getSignature());
        }
        
        @Test
        public void invalidAlgorithm() throws JSONException, MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);

            final byte[] envelopeBytes = envelope.getBytes();
            final JSONObject jo = new JSONObject(new String(envelopeBytes));
            jo.put(KEY_ALGORITHM, "x");
            
            final byte[] joJson = jo.toString().getBytes();
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(joJson);
            assertNull(joEnvelope.getAlgorithm());
            assertArrayEquals(joJson, joEnvelope.getSignature());
        }
        
        @Test
        public void missingSignature() throws JSONException, MslCryptoException, MslEncodingException {
            final MslSignatureEnvelope envelope = new MslSignatureEnvelope(algorithm, SIGNATURE);

            final byte[] envelopeBytes = envelope.getBytes();
            final JSONObject jo = new JSONObject(new String(envelopeBytes));
            jo.remove(KEY_SIGNATURE);
            
            final byte[] joJson = jo.toString().getBytes();
            final MslSignatureEnvelope joEnvelope = MslSignatureEnvelope.parse(joJson);
            assertNull(joEnvelope.getAlgorithm());
            assertArrayEquals(joJson, joEnvelope.getSignature());
        }
    }
}
